catalyst-gateway/bin/src/service/api/documents/common/mod.rs

```diff
@@ -0,0 +1,125 @@
+    // Find the doc in the static templates first
+    if let Some(doc) = get_doc_static_template(document_id) {
+        return Ok(doc);
+    }
+
+    // If doesn't exist in the static templates, try to find it in the database
+    let db_doc = FullSignedDoc::retrieve(document_id, version).await?;
+    db_doc.raw().try_into()
+}
+
+/// A struct which implements a
+/// `catalyst_signed_doc::providers::CatalystSignedDocumentProvider` trait
+pub(crate) struct DocProvider;
+
+impl catalyst_signed_doc::providers::CatalystSignedDocumentProvider for DocProvider {
+    async fn try_get_doc(
+        &self, doc_ref: &catalyst_signed_doc::DocumentRef,
+    ) -> anyhow::Result<Option<CatalystSignedDocument>> {
+        let id = doc_ref.id.uuid();
+        let ver = doc_ref.ver.map(|uuid| uuid.uuid());
+        match get_document(&id, ver.as_ref()).await {
+            Ok(doc) => Ok(Some(doc)),
+            Err(err) if err.is::<NotFoundError>() => Ok(None),
+            Err(err) => Err(err),
+        }
+    }
+}
+
+// TODO: make the struct to support multi sigs validation
+/// A struct which implements a
+/// `catalyst_signed_doc::providers::CatalystSignedDocumentProvider` trait
+pub(crate) struct VerifyingKeyProvider(
+    Vec<(catalyst_signed_doc::IdUri, ed25519_dalek::VerifyingKey)>,
+);
+
+impl catalyst_signed_doc::providers::VerifyingKeyProvider for VerifyingKeyProvider {
+    async fn try_get_key(
+        &self, kid: &catalyst_signed_doc::IdUri,
+    ) -> anyhow::Result<Option<ed25519_dalek::VerifyingKey>> {
+        let res = self.0.iter().find(|(id, ..)| id == kid).map(|item| item.1);
+
+        Ok(res)
+    }
+}
+
+impl VerifyingKeyProvider {
+    /// Attempts to construct an instance of `Self` by validating and resolving a list of
+    /// Catalyst Document KIDs against a provided RBAC token.
+    ///
+    /// This method performs the following steps:
+    /// 1. Verifies that **all** provided `kids` match the Catalyst ID from the RBAC
+    ///    token.
+    /// 2. Extracts the role index and rotation from each KID.
+    /// 3. Retrieves the latest signing public key and rotation state associated with the
+    ///    role for each KID from the registration chain.
+    /// 4. Collects and returns a vector of tuples containing the KID, signing key.
+    ///
+    /// # Errors
+    ///
+    /// Returns an `anyhow::Error` if:
+    /// - Any KID's short Catalyst ID does not match the one in the token.
+    /// - Indexed registration queries or chain building fail.
+    /// - The KID's role index and rotation parsing fails.
+    /// - The KID is not a singing key.
+    /// - The latest signing key for a required role cannot be found.
+    /// - The KID is not using the latest rotation.
+    pub(crate) fn try_from_kids(
+        token: &CatalystRBACTokenV1, kids: &[catalyst_signed_doc::IdUri],
+    ) -> anyhow::Result<Self> {
+        // validate rbac token and document KIDs (ignoring the role/rotation)
+        if kids
```

Comment:
Re: https://github.com/input-output-hk/catalyst-voices/pull/2038#discussion_r2031105502

Should expect only one kid